core/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -780,6 +780,17 @@ object IO {
     case x :: xs => x.fork *> forkAll(xs)
   }
 
+  /**
+   * Races a traversable collection of `IO[E, A]` against each other. If all of
+   * them fail, the last error is returned.
+   *
+   * _Note_: if the collection is empty, there is no action that can either
+   * succeed or fail. Therefore, the only possible output is an IO action
+   * that never terminates.
+   */
+  def raceAll[E, A](t: TraversableOnce[IO[E, A]]): IO[E, A] =
+    t.foldLeft(IO.never[E, A])(_ race _)
```

Comment:
I'm not convinced by how empty collections are handled. What about if we return `Option[IO[E, A]]`?
Or even better, we don't accept empty collections with a poor man's NEL like:

```scala
def raceAll[E, A](io: IO[E, A], ios: TraversableOnce[IO[E, A]]): IO[E, A] =
```

WDYT @jdegoes ? did you think of that case when defining the method signature?

Reply:
`never` has a certain elegance that is definitely alluring. If you race nothing, then of course it takes forever to complete.

I just wonder if developers will be confused or not by it. 

If we have a law:

`race(io, never) == io`

(Which we should and will)

Then we should have:

`raceAll(ios) == raceAll(never :: ios)`

which makes sense and implies:

`raceAll(Nil) == raceAll(never)`

Reply:
Isn't `IO.never` the same thing as `None`, given how it is defined?

Here was my reasoning: We can draw the parallel with finding the maximum in a collection of numbers. The only solution for an empty collection is -∞.

I think your updated signature makes sense and I'm happy to update it that way in order to get rid of the empty use case.

Reply:
@regiskuckaertz I was arguing that your existing implementation is correct and does not need changing.

However, we could always add an additional `def raceAll1` that uses the above signature (`head: IO[E, A], tail: TraversableOnce[IO[E, A]]`).

Reply:
My bad, saw your message just after hitting "Comment". It may be that `IO[E, A]` with `race` and `never` forms a commutative monoid. I'll add `raceAll1` 👍

Reply:
Sure, I'm not talking about anything wrong in the implementation, it is nice. But could lead to bugs hard to find.